### PR TITLE
Admin update

### PR DESCRIPTION
**Pull request by stevewh**
Add file structure to /admin for various modules/libraries and examples

Created a first version of a READ statistics library and moved to admin/module subdirectory

Created a first version of a links maintenance library and moved to admin/module subdirectory

moved readQueryCursor.py to admin/module subdirectory

### CODE REVIEW

**Comment by stevewh** [comment]
Just added a .gitignore file to each output subdir   with  a single line containing *
This will make sure that git ignores all files added to output and will also ensure the directory get created.

File: admin/data_tools/text_edition_cleanup.ipynb [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+    "#add parent dir to module search path to find readQueryCursor module\n",
+    "sys.path.insert(0,os.path.dirname(os.getcwd()))\n",
+    "#print(sys.path)\n",
+    "import readQueryCursor as rqc\n",
+    "\n",
+    "\n",
+    "dbname = 'homer3'\n",
+    "\n",
+    "conf = rqc.readConnConfig\n",
+    "conf['database'] = dbname\n",
```
Comment (selaudin) [inline admin/data_tools/text_edition_cleanup.ipynb:31]: Is it a good practice to also hide the credentials even in python notebooks?

File: docker/pg/Dockerfile [hunk 2]
```diff
@@ -1,4 +1,4 @@
-FROM postgres:10.14
+FROM postgres:10
```
Comment (selaudin) [inline docker/pg/Dockerfile:1]: In the future we should upgrade the postgres version, since version 10 it is supported until November 2022

**Review by selaudin — COMMENTED** [review]
I have went through the code changes, everything looks fine! (I have not tested them)

**Comment by stevewh** [comment]
I would say yes!  Can you suggest a change?

File: admin/data_tools/text_edition_cleanup.ipynb [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+    "conf['user'] = 'postgres'\n",
+    "conf['password'] = 'gandhari'\n",
```
Comment (selaudin) [inline admin/data_tools/text_edition_cleanup.ipynb]:
```suggestion
    "from dotenv import load_dotenv\n",
    "\n",
    "\n",
    "load_dotenv()\n",
    "\n",
    "dbname = 'homer3'\n",
    "\n",
    "conf = rqc.readConnConfig\n",
    "conf['database'] = dbname\n",
    "conf['user'] = os.getenv('DB_USERNAME')\n",
    "conf['password'] = os.getenv('DB_PASSWORD')\n",
```

**Comment by stevewh** [comment]
Nice. Can you write a .env readme or md as to how this gets set up?
tell where the .env needs to be and what to put in it. Perhaps we should
add a sampleDotEnv.txt file somewhere.

Also perhaps suggest how you would create a single source of this information that would also feed the .yml file
      POSTGRES_USER: adminusername
      POSTGRES_PASSWORD: password

File: docker/pg/Dockerfile (hunk 2 above)
Comment (stevewh) [reply docker/pg/Dockerfile:1]: Going to ver 12 will require some changes to the auth code. We should move forward with changing the current postgresql abstime type usage to timestamp before we move to 12+.  We will also need to determine the best migration path for existing databases.

File: admin/data_tools/text_edition_cleanup.ipynb (hunk 3 above)
Comment (stevewh) [reply admin/data_tools/text_edition_cleanup.ipynb]: did you change requirements.txt to include dotenv?

I think using the env_file from docker-compose.yml loads the environment. If so, perhaps we don't need load_dotenv().

Comment (selaudin) [reply admin/data_tools/text_edition_cleanup.ipynb]: I have, in addition to pandas(to read csv file), requests(http request for the image), pillow(to open the image and get the size) and livereload(for web development, a server which in our case, helps to always get the last updated file from the file directory, not from the cached version stored in browser).

Yes, I also think if we import the .env file to docker-compose.yml, we would not have to use load_dotenv() to load in the Python environment. (I would have to check this)

**Comment by stevewh** [comment]
I made the changes to the files and committed them to this PR branch. Have
a look.

**Comment by selaudin** [comment]
'**admin/ReadCursorSample.py'**, line 16-17, user and password should be given default name like in sample.env.

File: admin/module/readLinkMaintenance.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        return []
+    prefix,entID = containerGID.split(":")
+    useGIDForm = False
+    if not prefix in linkColTableLookupByPrefix.keys():
+        print(f'call to getChildLinkdsForContainer unexpected prefix "{prefix}"')
+        return []
+    else:
+        tblName, colName = linkColTableLookupByPrefix.get(prefix)
+        myRQC.query(f"select {colName} from {tblName} where {prefix}_id = {entID};")
+        return myRQC.getRow(0)[0]
```
Comment (selaudin) [inline admin/module/readLinkMaintenance.py]: Maybe I am missing, but I don't understand how you can query like this. Shouldn't you at least pass a ReadQueryCursor parameter in the function?

File: admin/requirements.txt [hunk 5]
```diff
@@ -19,7 +19,10 @@ jinja2>=2.11.3
 jupyter-client==6.1.7
 jupyter-core==4.7.0
 MarkupSafe==1.1.1
+matplotlib
 mccabe==0.6.1
+numpy
+pandas
 parso==0.7.1
 pickleshare==0.7.5
 prompt-toolkit==3.0.8
```
Comment (selaudin) [inline admin/requirements.txt:28]:
```suggestion
prompt-toolkit==3.0.8
python-dotenv
```

For loading the environment variables when working locally in python notebook!

File: sample.env [hunk 6]
```diff
@@ -0,0 +1,5 @@
+POSTGRES_DB=postgres
+POSTGRES_USER=postgres
+POSTGRES_PASSWORD=adminpwd
+PGADMIN_DEFAULT_EMAIL=xx@xx.uu.com
+PGADMIN_DEFAULT_PASSWORD=defaultuserpwd
```
Comment (selaudin) [inline sample.env:5]: I would have to add later SERVERNAME=localhost, for telling COCO if the code is being run locally or in the server.

**Review by selaudin — COMMENTED** [review]
Everything looks fine, except that graCountsByTxtInv, was taking too long. And also I would suggest that in notebooks **from dotenv** to **import load_dotenv** and call **load_dotenv()**, to load the .env vars locally. Otherwise then if we would define a cell like
`
import os, sys
import numpy as np
#add parent dir to module search path to find readQueryCursor module
sys.path.append(os.path.dirname(os.getcwd()))
import module.readQueryCursor as rqc

myRQC = rqc.ReadQueryCursor(conf={
             'host':'localhost',
             'port':'5432',
             'database':'dbName',
             'user': 'dbUser',
             'password':'dbPassword'
            })
`, so developers can config it directly there. Either one works for me!

File: admin/module/readLinkMaintenance.py (hunk 4 above)
Comment (stevewh) [reply admin/module/readLinkMaintenance.py]: Correct, good catch this is a work in progress and should not have been committed yet. I will remove it.

File: sample.env (hunk 6 above)
Comment (stevewh) [reply sample.env:5]: The servername is set in the Dockerfile for READ (php-apache/Dockerfile 
 && echo "ServerName localhost" >> /etc/apache2/apache2.conf

Perhaps there is a way to get this from the request headers?

Comment (selaudin) [reply sample.env:5]: It is not that I need to acquire that server runs on localhost, but a variable like APP_ENV=server or APP_ENV=local

**Review by selaudin — CHANGES_REQUESTED** [review]
In **admin/ReadCursorSample.ipynb**, please don't forget to remove the credentials.
Also in **read/.gitignore** can you add the following lines:

`
.idea

.ipynb_checkpoints

.env
`

File: admin/data_tools/text_edition_cleanup.ipynb (hunk 3 above)
Comment (stevewh) [reply admin/data_tools/text_edition_cleanup.ipynb]: Anyone using the code locally would have to ensure that the environment gets loaded or add the information directly.

**Review by stevewh — COMMENTED** [review]
Moving forward with merge

**Pull request merged**